charms/reactive/altrelations.py

```diff
@@ -41,9 +44,68 @@ def from_name(cls, relation_name):
         return cls(relation_name)
 
     @classmethod
-    def from_state(cls, state):
-        value = get_state(state)
+    def from_flag(cls, flag):
+        value = _get_flag_value(flag)
         if value is None:
             return None
         relation_name = value['relation']
         return cls.from_name(relation_name)
+
+
+class Endpoint(RelationFactory):
+    @classmethod
+    def _startup(cls):
+        for endpoint in hookenv.relation_types():
+            rids = hookenv.relation_ids(endpoint)
+
+            # populate context based on attached relations
+            relf = relation_factory(endpoint)
+            if issubclass(relf, cls):
+                if rids:
+                    rel = relf(endpoint)
+                    setattr(context.endpoints, endpoint, rel)
+                else:
+                    setattr(context.endpoints, endpoint, None)
+
+            # manage automatic relation flags
+            toggle_flag('relations.{}.joined'.format(endpoint), rids)
+            for rid in rids:
+                for unit in hookenv.related_units(rid):
+                    rel_data = hookenv.relation_get(unit=unit, rid=rid)
+                    for key, value in rel_data.items():
+                        data_key = 'relations.{}.{}.{}.{}'.format(
+                            endpoint, rid, unit, key)
+                        if data_changed(data_key, value):
+                            set_flag('relations.{}.changed'.format(endpoint))
+                            set_flag('relations.{}.changed.{}'.format(
+                                endpoint, key))
+
+    def __init__(self, relation_name):
+        self.relation_name = relation_name
+
+    def set_flag(self, flag):
+        """
+        Set flag, tied to this relation.
+
+        Usage:
+
+            @when('relations.{relation_name}.changed')
+            def changed(self):
+                if any('host' in unit.receive for unit in self.units):
+                    self.set_flag('relations.{relation_name}.available')
+        """
+        set_flag(flag.format(relation_name=self.relation_name))
+
+    @classmethod
+    def from_name(cls, relation_name):
+        return cls(relation_name)
+
+    @classmethod
+    def from_flag(cls, flag):
+        parts = flag.split('.')
+        if len(parts) < 3 or parts[0] != 'relations':
+            return None
+        return cls.from_name(parts[1])
+
+
+hookenv.atstart(Endpoint._startup)
```

Comment:
Endpoint._startup should not be called in 'restricted' mode, or metrics and similar hooks are going to become heavyweight.

Reply:
`_run_atstart` isn't called in restricted mode, so it won't be.